### ISSUE TITLE
Incorrect bias with large regularization; difference with R's glmnet

### ISSUE THREAD

**Issue by aribenjamin** [glm-tools/pyglmnet#183]
__Short version:__ pyglmnet doesn't work with high regularization but a GLM in Keras does. Both give negative psuedo-R2 scores for all neurons in the S1 set. R's glmnet gives positive scores, and pyglmnet gives different results when running on R's regularization path.
___
A GLM should be able to predict the mean with high regularization, and find a pseudo-R2 value of 0. I'm not able to get pyglmnet to do this on an S1 dataset. Consider

```python
X = data[['handPos_x','handPos_y','handVel_x','handVel_y']].values

neuron_n = 15
y = math['spikes'][neuron_n]

def glm_pyglmnet(Xr, Yr, Xt):
    glm = GLM(distr='poisson', alpha=0.1, reg_lambda=10, tol=1e-3, 
        learning_rate=0.01, max_iter=10000, eta=4.0, random_state=1, verbose=False,
            )
    glm.fit(Xr, Yr)
    Yt = glm.predict(Xt)[0]
    
    print("\nMean y_test: "+str(np.mean(Yr))+"\n GLM thinks mean y: "+str(np.mean(glm.ynull_))+
      "\n exp(beta0) is "+str(np.exp(glm.fit_[0]['beta0'][0]))+"\n mean yhat is "+str(np.mean(Yt))+"\n" )

    return Yt

Yt_hat, PR2 = fit_cv(X, y, 'glm_pyglmnet', n_cv=4, verbose=2, label=[])
```
with output
```
...runnning cv-fold 1 of 4

Mean y_test: 0.878404521248
 GLM thinks mean y: 0.878404521248
 exp(beta0) is 1.48575497076e-74
 mean yhat is 1.48575497076e-74

pR2:  -41.6519743746
```
The output looks like this, with smoothed y in orange and smoothed yhat in blue. 
![image](https://cloud.githubusercontent.com/assets/9724569/21241718/94a204fc-c2d6-11e6-996a-ae88e89795eb.png)

I've played extensively with the learning rate, alpha, eta, iterations, a `score_metric` of `psuedo_R2`. I get similar poor results for the default regularization path.

A strange clue is that I get similar results when I pass `X = np.ones((len(y),))` as input. This indicates to me that something is funky with the beta0 optimization.
___
A GLM in Keras can better learn the mean with similar regularization:
```python
def glm_keras(Xr, Yr, Xt,l1=0.1,l2=10):
    model = Sequential()
    model.add(Dense(1, input_dim=np.shape(Xr)[1], init='uniform',W_regularizer=l1l2(l1,l2), activation='linear'))
    model.add(Lambda(lambda x: np.exp(x)))
    model.compile(loss='poisson', optimizer='rmsprop')
    model.fit(Xr, Yr, nb_epoch=4, verbose=0, validation_split=0.0)
    Yt = model.predict_proba(Xt, verbose=0)
    return Yt[:,0]

Yt_hat, PR2 = fit_cv(X, y, 'glm_keras', n_cv=4, verbose=2, label=[])
```
out:
```
pR2_cv: -0.009022 (+/- 0.002800)
Mean y: 0.879214215846, mean yhat: 0.866071888691
```

__But here's the kicker__: R's glmnet can consistently find positive pR2 scores:
```python
import rpy2.robjects as ro
import rpy2.robjects.numpy2ri as n2r
n2r.activate()

r = ro.r
r.library('glmnet')


def r_glmnet(Xr,Yr,Xt):
    yr = ro.FloatVector(Yr) # use factors
    trained_model = r['glmnet'](Xr, yr, family="poisson",alpha = 0.1)
    lambda_min = np.asanyarray(trained_model.rx2('lambda'))[-1]
    Yt = r['predict'](trained_model,s=lambda_min,newx=Xt,type="response")
    return np.array(list(Yt))

Yt_hat, PR2 = fit_cv(X, y, 'r_glmnet', n_cv=4, verbose=2, label=[])
```
out:
```
pR2_cv: 0.000903 (+/- 0.000120)
Mean y: 0.879214215846, mean yhat: 0.879137023226
```
R defines its own regularization path, and so perhaps this just a better choice. This isn't the case, though - when I feed R's choice of regularization path to pyglmnet it gets different results.
```python
path = np.asanyarray(trained_model.rx2('lambda'))

def glm_pyglmnet(Xr, Yr, Xt):
    glm = GLM(distr='poisson', alpha=0.1, reg_lambda=path, tol=1e-3)
    glm.fit(Xr, Yr)
    Yt = glm.predict(Xt)[0]
    
    print("\nMean y_test: "+str(np.mean(Yr))+"\n GLM thinks mean y: "+str(np.mean(glm.ynull_))+
      "\n exp(beta0) is "+str(np.exp(glm.fit_[0]['beta0'][0]))+"\n mean yhat is "+str(np.mean(Yt))+"\n" )

    return Yt

Yt_hat, PR2 = fit_cv(X, y, 'glm_pyglmnet', n_cv=4, verbose=2, label=[])
```
yields:

```
...runnning cv-fold 1 of 4

Mean y_test: 0.885599391371
 GLM thinks mean y: 0.885599391371
 exp(beta0) is 0.000411927287626
 mean yhat is 6.24597208743

pR2:  -17.4722837027
...runnning cv-fold 2 of 4

Mean y_test: 0.879860884686
 GLM thinks mean y: 0.879860884686
 exp(beta0) is 0.000364188044748
 mean yhat is 7.88373870267

pR2:  -19.783277945
```
I have very little idea what's going on here; perhaps one of you all can make better sense of it. It could very well be two different issues. I've uploaded this notebook to  [https://github.com/aribenjamin/S1_GLMs](https://github.com/aribenjamin/S1_GLMs)  if someone wants the playground.

**Comment by aribenjamin** [comment]
This is resolved! The key error was that, above,

```python
Yt = glm.predict(Xt)[0]
```
should be
```python
Yt = glm[-1].predict(Xt)
```
___
I learned a few lessons diagnosing this.
1. The results are very sensitive to `tol`, ` learning_rate`, and `max_iter`. They should be optimized on any implementation.
2. `eta=2' is the default value and works well.
3. Initializing the value of all `beta`s doesn't affect the result too much, even when initialized to the final value.
4. Don't assign any variables to the name `math`. Odd and hard-to-diagnose errors result.

Thanks to Pavan for the help here.
